cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/service/BusinessMetadataStore.java

```diff
@@ -0,0 +1,191 @@
+import co.cask.cdap.common.ServiceUnavailableException;
+import co.cask.cdap.data2.datafabric.dataset.DatasetsUtil;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.DatasetManagementException;
+import co.cask.cdap.data2.dataset2.tx.Transactional;
+import co.cask.cdap.data2.metadata.dataset.BusinessMetadataDataset;
+import co.cask.cdap.proto.Id;
+import co.cask.tephra.TransactionExecutor;
+import co.cask.tephra.TransactionExecutorFactory;
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+import com.google.common.base.Supplier;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Iterators;
+import com.google.inject.Inject;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * Implements operations on {@link BusinessMetadataDataset} transactionally.
+ */
+public class BusinessMetadataStore {
+
+  private static final Id.DatasetInstance BUSINESS_METADATA_INSTANCE_ID =
+    Id.DatasetInstance.from(Id.Namespace.SYSTEM, "business.metadata");
+  private final Transactional<BusinessMdsIterable, BusinessMetadataDataset> txnl;
+
+  @Inject
+  public BusinessMetadataStore(TransactionExecutorFactory txExecutorFactory, final DatasetFramework dsFramework) {
+    this.txnl = Transactional.of(txExecutorFactory, new Supplier<BusinessMdsIterable>() {
+      @Override
+      public BusinessMdsIterable get() {
+        try {
+          Object dataset = DatasetsUtil.getOrCreateDataset(dsFramework, BUSINESS_METADATA_INSTANCE_ID,
+                                                            BusinessMetadataDataset.class.getSimpleName(),
+                                                            DatasetProperties.EMPTY, null, null);
+          if (dataset instanceof BusinessMetadataDataset) {
+            return new BusinessMdsIterable((BusinessMetadataDataset) dataset);
+          }
+          return new BusinessMdsIterable(new BusinessMetadataDataset((IndexedTable) dataset));
+        } catch (DatasetManagementException | IOException | ServiceUnavailableException e) {
+          throw Throwables.propagate(e);
+        }
+      }
+    });
+  }
+
+  /**
+   * Adds/updates metadata for the specified {@link Id.NamespacedId}.
+   *
+   * @param entityId the {@link Id.NamespacedId} for which metadata is to be updated
+   * @param metadata metadata (represented as a map) to update
+   */
+  void addMetadata(final Id.NamespacedId entityId, final Map<String, String> metadata) {
+    txnl.executeUnchecked(new TransactionExecutor.Function<BusinessMdsIterable, Void>() {
+      @Override
+      public Void apply(BusinessMdsIterable input) throws Exception {
+        for (Map.Entry<String, String> entry : metadata.entrySet()) {
+          input.businessMds.createBusinessMetadata(entityId, entry.getKey(), entry.getValue());
```

Comment:
What do we do here if user specifies `tags` as one of the key in `metadata` map?